add intersection behavior for RSP

xunpan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Add an entry to CHANGELOG.md if the PR represents a user-visible change.
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
We add intersection behavior because it makes sense to result placement for resource propagation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #981

**Special notes for your reviewer**:
- placement computing rule is get from current federated resource calculation. That means the related rule is reused in `RSP` for placement calculation 
- I'm trying to add test and add test cases. So, please review when related tasks are done.

xunpan commented:
/retest

xunpan commented:
/retest

xunpan commented:
/test

xunpan commented:
@irfanurrehman 
It is ready to review.

irfanurrehman commented:
Thanks for doing this @xunpan. I will positively take a look at this tomorrow.

irfanurrehman commented on pkg/schedulingtypes/replicascheduler.go:
```diff
@@ -190,6 +191,34 @@ func (s *ReplicaScheduler) Reconcile(obj pkgruntime.Object, qualifiedName ctluti
+	if rsp.Spec.PlacementIntersection {
+		klog.V(3).Infof("Computing placement of resource %q", qualifiedName)
+
+		fedClusters, err := s.podInformer.GetReadyClusters()
```
Is this not the same names that you already have at line 167 above (`clusterNames, err := s.clusterNames()`)?

xunpan replied:
you're right. I will update it.

irfanurrehman commented on charts/kubefed/charts/controllermanager/crds/crds.yaml:
```diff
@@ -860,6 +860,13 @@ spec:
                   provided. If omitted, clusters without explicit preferences should
                   not have any replicas scheduled.
                 type: object
+              placementIntersection:
```
I think `IntersectWithClusterSelector` will be a better name.

irfanurrehman commented on pkg/controller/util/placement.go:
```diff
@@ -86,3 +89,92 @@ func SetClusterNames(obj *unstructured.Unstructured, clusterNames []string) erro
+	clusterNames := placement.ClusterNames()
+	// Only use selector if clusters are nil. An empty list of
+	// clusters implies no clusters are selected.
+	if clusterNames == nil {
```
The problem with this approach is that RSP uses the clusternames itself to specify its decision, so the intersection of the cluster selector with the RSP decision will not work if both the cluster selector and the clusternames are specified on the target. 
Also, for the scenario where only the clusterselector is specified, the first reconcile result will be the intersection of the clusterselector and the RSP decision, but next time onwards the clusterselector will be ignored becaue RSP would have written clusternames on the target resource. So if the clusterselector is updated after that the update will be ignored.

irfanurrehman commented on pkg/schedulingtypes/plugin.go:
```diff
@@ -133,6 +143,24 @@ func (p *Plugin) FederatedTypeExists(key string) bool {
+	if p.typeConfig.GetNamespaced() {
+		return util.ComputeNamespacedPlacement(fedObject, fedNsObject, clusters, p.limitedScope)
+	}
+	return util.ComputePlacement(fedObject, clusters)
```
Please see my comment above about using this as is here. I think we need a separate function which gets only the clusterselector from the target resource. We are trying to intersect with the clusterselector only, not the clusternames.

irfanurrehman commented on pkg/schedulingtypes/replicascheduler.go:
```diff
@@ -190,6 +191,34 @@ func (s *ReplicaScheduler) Reconcile(obj pkgruntime.Object, qualifiedName ctluti
+			return ctlutil.StatusError
+		}
+
+		preferredClusters := []string{}
```
Could not understand the need of `preferredClusters` it seems to be a copy of `resultClusters`

irfanurrehman replied:
nit: @xunpan this is still unresolved but I will approve as this does not seem to harm the logic at all.

irfanurrehman commented on test/e2e/scheduling.go:
```diff
@@ -202,10 +230,21 @@ func createTestObjs(tl common.TestLogger, client genericclient.Client, typeConfi
+
+	if rspSpec.PlacementIntersection {
+		clusters = append(clusters, fedClusters[0])
+		err = util.SetClusterNames(fedObject, clusters)
```
I think this needs to be tested with a clusterselector rather then a cluster name.

irfanurrehman reviewed (changes requested):
@xunpan Many thanks for spending time on this.
However, I think there are some rework that would be needed on this PR. Please take a look at my comments.

xunpan commented on charts/kubefed/charts/controllermanager/crds/crds.yaml:
```diff
@@ -860,6 +860,13 @@ spec:
                   provided. If omitted, clusters without explicit preferences should
                   not have any replicas scheduled.
                 type: object
+              placementIntersection:
+                description: If set to true then the target kind (FederatedDeployment
+                  or FederatedReplicaset) placement will be intersected with Clusters
+                  of the RSP to determine the result clusters of the federated resource.
+                  if set false or not defined, Clusters in the RSP overriding the
+                  placment in the federated resource.
+                type: boolean
```
OK.

xunpan commented on pkg/apis/scheduling/v1alpha1/replicaschedulingpreference_types.go:
```diff
@@ -45,6 +45,14 @@ type ReplicaSchedulingPreferenceSpec struct {
+	// if set false or not defined, Clusters in the RSP overriding the placment in the federated
+	// resource.
+	// +optional
+	PlacementIntersection bool `json:"placementIntersection,omitempty"`
```
ClusterSelector is a little misleading I guess. For a federated resource, we have 
```
spec:
  placement:
    clusters:
        ...
    clusterSelector:
        ...
```
I use Placement here because it includes both `clusters` and `clusterSelector`.

Any comments?

hectorj2f reviewed (commented):
Could you add some documentation with an example as part of our current docs ?

hectorj2f commented:
@xunpan Could you address the requested changes by @irfanurrehman ?

xunpan commented:
Yes. The code was done and I'm testing it locally. The updated diff will be committed by this week.

irfanurrehman commented:
> Yes. The code was done and I'm testing it locally. The updated diff will be committed by this week.

@xunpan I missed checking on this. I see that you did some updates. Were they finalised by you? Should I check them again?

xunpan commented:
@irfanurrehman I forgot to send you a message after update the PR.

I think I addressed your comments above. One does not changed:  
[Q] `preferredClusters` it seems to be a copy of `resultClusters`
[A] It is used to transform `sets.String` to `string` slice

irfanurrehman commented:
Thanks for doing this @xunpan.
/lgtm
